engine/client/gl_rmain.c

```diff
@@ -1326,99 +1328,21 @@ void R_RenderFrame( const ref_params_t *fd, qboolean drawWorld )
 	GL_BackendEndFrame();
 }
 
-_inline void GL_InsertBlackFrame( void )
-{
-	// No strobing on the console
-	if( CL_IsInConsole() )
-	{
-		pglEnable( GL_SCISSOR_TEST );
-		pglScissor( con_rect.x, (-con_rect.y) - (con_rect.h*1.25), con_rect.w, con_rect.h ); // Preview strobe setting on static
-		pglClearColor( 0.0f, 0.0f, 0.0f, 1.0f );
-		pglClear( GL_COLOR_BUFFER_BIT );
-		pglDisable( GL_SCISSOR_TEST );
-	}
-	else
-	{
-		pglClearColor( 0.0f, 0.0f, 0.0f, 1.0f );
-		pglClear( GL_COLOR_BUFFER_BIT );
-	}
-}
-
-/*
-===============
-R_Strobe
-
-TODO: Consider vsync timings and do not render the supposed black frame at all.
-===============
-*/
-void R_Strobe( void )
-{
-	static int sCounter = 0;
-	int getInterval = r_strobe->integer; // Check through modified tag first?
-
-	if( CL_IsInMenu() )
-	{
-		R_Set2DMode(false);
-		return;
-	}
-
-	if( !getInterval || ( !gl_swapInterval->integer && getInterval ) )
-	{
-		if( getInterval ) //If v-sync is off, turn off strobing
-		{
-			Cvar_Set( "r_strobe", "0" );
-			MsgDev( D_WARN, "Strobing (Black Frame Replacement) requires V-Sync not being turned off! (gl_swapInterval != 0)\n" );
-		}
-		else if( sCounter )
-			sCounter = 0;
-
-		// flush any remaining 2D bits
-		R_Set2DMode(false);
-		return;
-	}
-	
-	// If interval is positive, insert (replace with) black frames.
-	// For example result of interval = 3 will be: "black-black-black-normal-black-black-black-normal-black-black-black-normal"
-	if( getInterval > 0 )
-	{
-		if( sCounter < getInterval )
-		{
-			GL_InsertBlackFrame();
-			++sCounter;
-		}
-		else
-		{
-			sCounter = 0;
-			R_Set2DMode( false );
-		}
-	}
-	// If interval is negative, the procedure will be the opposite reverse.
-	// For example result of interval = -4 will be: "normal-normal-normal-normal-black-normal-normal-normal-normal-black"
-	else
-	{
-		getInterval = abs( getInterval );
-		if( sCounter < getInterval )
-		{
-			++sCounter;
-			R_Set2DMode( false );
-		}
-		else
-		{
-			GL_InsertBlackFrame();
-			sCounter = 0;
-		}
-	}
-}
-
 /*
 ===============
 R_EndFrame
 ===============
 */
 void R_EndFrame( void )
 {
-	R_Strobe();
+#ifdef STROBE_ENABLED
+	Strobe_Invoker((void**)(&STROBE_CORE), STROBE_CORE_EXPORTEDFUNC_constructor, STROBE_CORE_EXPORTEDFUNC_main, STROBE_CORE_EXPORTEDFUNC_destructor);
```

Comment:
Now it looks more ugly. Can you just look to other parts of engine and do similar?

Reply:
Yeah I know it is very ugly. But I do not think that this can be not ugly with the object orient structure intact. It is C after all.

I have thought using GObject or something like that but decided not to use because I want this strobe implementation to be self-contained.

Reply:
In what parts of the engine for instance ?

Reply:
* basecmd.c: everything is in base_cmd.c and only minimal possible functions are visible outside in base_cmd.h.
* masterlist.c: there is only three functions are visible outside. Internal states are defined in global structure(encapsulated).
* ImageLib & SoundLib as an example of images or sound manager in engine.
* joyinput.c: abstraction over platform-dependent gamepad API.
* avikit.c: opaque movie_state_t visible outside. You just operate with simple API. 
* cfgscript.c, identification.c, et cetera. Just everything in engine is structured.
* Graphics rendering is complex and stays in it's own namespace for files: gl_*.c. The same for client, server and sound.
* All functions are named in the such scheme: <namespace>_<func>, so there is almost no confusion. Cvar_Get means that this function is from "Cvar" subsystem. HTTP_Run shows that this is something for HTTP support in engine.

You will not find any feature, that's implementation is smudged over the whole engine. Only in such way engine development feels friendly and comfortable. Thanks to Carmack for it. Look into Quake engine, this is fully dervied from it. Some Quake-derived engines still use this, like QFusion, that's why I found QFusion code is beatiful. 

I don't know why you think that you need OOP here. Yes, you may simulate some aspects from it, like encapsulation(example above) or polymorphism(see in backends. :)). And that's enough for everything in game engine.

Reply:
Don't think that it's too hard to send patches for such small and young(compared to original Xash3D) project. I just want to keep single and simple way to work with engine source code. Otherwise it will go fast into the pile of garbage of too different code, like Darkplaces.

Reply:
But the concept here is different. This way a developer can develop their own strobe implementation without altering the core implementation and run it along with different implementations. Also with this way, even if STROBE_ENABLED is defined, strobe will not allocate unnecessary space in memory unless user sets r_strobe to non zero (for core implementation).

https://vgy.me/xnhxKP.png . I think if oop is not used here, either "strobe" will be all over the place or it will suck really hard memory-wise. Also maintenancing new implementations will be much harder.

I think it is a fair trade to use oop here.

Reply:
This a generic game engine in first place. If someone want to develop their strobe algorithm, they can implement it in client library.

Reply:
Good idea!  Long term (eventually) this could be a generic strobe library.   

Problem is that strobing requires very low-level synchronization and requires a certain amount of very tight 2-way integration as fuzun's diagram explains -- https://vgy.me/xnhxKP.png -- very good points.   The strobe API should ideally have its own independently configurable features (e.g. black frame ratios) and maybe 'hint' to the rest of the game of its preferred patterns (e.g. a "biasing" to 72fps during 144Hz operation, or 45fps/60fps/90fps modes for various strobe patterns of 180Hz strobing -- the game has to change its framerate to adapt to the preferred strobe algorithm -- necessitiating a 2-way influence between the strobe API and the game engine, to maximize smoothness and minimize input lag, etc)

It's a code architectural challenge -- how to do it in a clean way, is a super-big-challenge. One can only do their best to self-contain the strobing, but will require quite a bunch of low-level hooks sprinkled throughout any game engine or emulators.   

Theoretical simplifications can be done such as adding detection of VBlanks (e.g. Direct3D9 RasterStatus.InVBlank but that is not always reliable) but it can be very hard to detect if a framebuffer flip (Present) made it to the correct refresh cycle.

Reply:
P.S. I see source code license says "This project uses Quake's C code style convention."

If there's a disagreement about using OOP or not .... how hard is it to use a client library approach?  Keep the OOP inside the library and simply use library calls from the favourite C project.   Sometimes larger amounts of prettier code is preferable over smaller amounts of uglier code -- even from a maintainability perspective.  Beauty is in the eye of the beholder...

I'm not sure how hard this is; just brainstorming ways to make everyone happy :)

Reply:
@mdrejhon - Thank you for sharing opinion. I want to clarify a few points.

"it's a good idea to minimize amount of code outside the strobe module the absolute bare minimum where possible, and blackbox as much stuff as possible into the strobe library" this was the main point of making this built on object oriented structure so that other will be able to made with ease and with good memory handling. But it also stands out among the other Xash3d source files (structure is fundementally different as @a1batross have pointed out) and because it is OOP in C.

With this way, a different implementation can be made really easily. I think that strobing is a very complex thing and the "core" implementation itself is not sufficient. For example it does not support varied positive phase shifting (think as PWM signal, not the other phase).

Look at the template files as an example (They have a few errors though that I will correct), they contain the basic structure for making a new strobing implementation. It is very easy to make because Strobe API provides all necessary backend functions. Even the debug script is generated by that.

The only thing that needs to be done is implementing the algorithm and updating protected members of the abstract base class/structure: Strobe-api through the new code. And including the new .h file and calling Strobe_Invoke function each frame and/or calling (own) debughandler function for drawing the debug script.

And yeah beyond this point it is all about decisions. Since it is C, I do not think that it will get much better than this. If it will have less OOP, maintenance and implementation will be harder. If it will have more OOP, it will be more bloated. There is a trade.